File: empyrean-py/python/empyrean/math.py
"""Covariance and mixture math primitives.

Public Python wrappers for the 6-dimensional state-covariance utilities
exposed by the engine: dominant-eigenvector extraction (used wherever
you need the principal axis of an uncertainty ellipsoid) and Gaussian
mixture splitting along that axis (the building block of the AGM
uncertainty propagation method).
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np


@dataclass(frozen=True)
class MixtureComponent:
    """One weighted component of a 6D Gaussian mixture.

    Returned by :func:`split_gaussian`.

    Attributes
    ----------
    weight : float
        Component weight (sum across components is 1.0).
    mean : numpy.ndarray
        6-element mean vector.
    covariance : numpy.ndarray
        6 × 6 covariance matrix.
    """

    weight: float
    mean: np.ndarray
    covariance: np.ndarray


def eigenvector_max_6x6(matrix: np.ndarray) -> tuple[float, np.ndarray]:
    """Dominant eigenvalue / eigenvector of a 6 × 6 symmetric matrix.

    Parameters
    ----------
    matrix : numpy.ndarray
        6 × 6 symmetric (covariance-like) matrix.

    Returns
    -------
    eigenvalue : float
        Largest eigenvalue.
    eigenvector : numpy.ndarray
        Corresponding 6-element unit eigenvector.
    """
    from empyrean._empyrean_rs import _eigenvector_max_6x6

    arr = np.ascontiguousarray(matrix, dtype=np.float64)
    if arr.shape != (6, 6):
        raise ValueError(f"matrix must be 6x6, got shape {arr.shape}")
    eigenvalue, eigenvector = _eigenvector_max_6x6(arr)
    return float(eigenvalue), np.asarray(eigenvector, dtype=np.float64)


def split_gaussian(
    mean: np.ndarray,
    covariance: np.ndarray,
    k: int,
) -> list[MixtureComponent]:
    """Split a 6D Gaussian into ``k`` weighted components along the
    dominant eigenvector of the covariance.

    The split direction is the principal axis of the input covariance
    (matches the engine's adaptive Gaussian-mixture splitter).

    Parameters
    ----------
    mean : numpy.ndarray
        6-element mean vector of the input distribution.
    covariance : numpy.ndarray
        6 × 6 covariance of the input distribution.
    k : int
        Number of mixture components. Must be ≥ 1.

    Returns
    -------
    list[MixtureComponent]
        ``k`` weighted components whose marginal sums to the input
        Gaussian.
    """
    from empyrean._empyrean_rs import _split_gaussian

    if k < 1:
        raise ValueError(f"k must be >= 1, got {k}")

    mean_arr = np.ascontiguousarray(mean, dtype=np.float64)
    cov_arr = np.ascontiguousarray(covariance, dtype=np.float64)
    if mean_arr.shape != (6,):
        raise ValueError(f"mean must have shape (6,), got {mean_arr.shape}")
    if cov_arr.shape != (6, 6):
        raise ValueError(f"covariance must be 6x6, got shape {cov_arr.shape}")

    result = _split_gaussian(mean_arr, cov_arr, k)
    weights = np.asarray(result["weights"], dtype=np.float64)
    means = np.asarray(result["means"], dtype=np.float64)
    covs = np.asarray(result["covariances"], dtype=np.float64)

    return [
        MixtureComponent(
            weight=float(weights[i]),
            mean=means[i].copy(),
            covariance=covs[i].copy(),
        )
        for i in range(k)
    ]

File: empyrean/src/math.rs
//! Covariance and uncertainty math utilities.

use crate::error::{Error, Result};

/// Compute the largest eigenvalue and corresponding eigenvector of a
/// 6×6 symmetric matrix.
pub fn eigenvector_max_6x6(matrix: &[[f64; 6]; 6]) -> Result<(f64, [f64; 6])> {
    let mut eigenvalue: f64 = 0.0;
    let mut eigenvector: [f64; 6] = [0.0; 6];
    let code = unsafe {
        empyrean_sys::empyrean_eigenvector_max_6x6(matrix, &mut eigenvalue, &mut eigenvector)
    };
    if code != 0 {
        return Err(Error::capture(code));
    }
    Ok((eigenvalue, eigenvector))
}

/// A component of a Gaussian mixture produced by [`split_gaussian`].
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct MixtureComponent {
    /// Component weight (sums to 1 across the mixture).
    pub weight: f64,
    /// Component mean (6-element state vector).
    pub mean: [f64; 6],
    /// Component 6x6 covariance.
    pub covariance: [[f64; 6]; 6],
}

/// Split a Gaussian (mean, covariance) into a mixture of `k` smaller
/// Gaussians along the eigenvector of maximum variance.
///
/// Returns a mixture of `k` components whose weighted moments match the
/// input mean and covariance to second order.
pub fn split_gaussian(
    mean: &[f64; 6],
    covariance: &[[f64; 6]; 6],
    k: usize,
) -> Result<Vec<MixtureComponent>> {
    let mut weights: Vec<f64> = vec![0.0; k];
    let mut means: Vec<[f64; 6]> = vec![[0.0; 6]; k];
    let mut covs: Vec<[[f64; 6]; 6]> = vec![[[0.0; 6]; 6]; k];

    let code = unsafe {
        empyrean_sys::empyrean_split_gaussian(
            mean,
            covariance,
            k,
            weights.as_mut_ptr(),
            means.as_mut_ptr(),
            covs.as_mut_ptr(),
        )
    };
    if code != 0 {
        return Err(Error::capture(code));
    }

    Ok((0..k)
        .map(|i| MixtureComponent {
            weight: weights[i],
            mean: means[i],
            covariance: covs[i],
        })
        .collect())
}
